lib/status.js

```diff
@@ -9,45 +10,101 @@ const CIRCUIT_BREAKER = Symbol('circuit-breaker');
+    this[STATS_WINDOW] = [];
+
+    function increment (status, property) {
+      return () => status[STATS_WINDOW][0][property]++;
+    }
+
+    circuit.on('success', increment(this, 'successes'));
+    circuit.on('failure', increment(this, 'failures'));
+    circuit.on('fallback', increment(this, 'fallbacks'));
+    circuit.on('timeout', increment(this, 'timeouts'));
+    circuit.on('fire', increment(this, 'fires'));
+    circuit.on('reject', increment(this, 'rejects'));
+
+    // Set up our statistical rolling window
+    const buckets = circuit.options.rollingCountBuckets;
+    const timeout = circuit.options.rollingCountTimeout;
+
+    // Add the first bucket to the window
+    this[STATS_WINDOW].unshift(stats());
+
+    // TODO: do we guard against divide by zero, and for
+    // greater accuracy, do we require that timeout be
+    // evenly divisible by the number of buckets?
+    const bucketInterval = Math.floor(timeout / buckets);
+    const interval = setInterval(() => {
+      if (this[STATS_WINDOW].length === buckets) {
+        this[STATS_WINDOW].pop();
+      }
+      this[STATS_WINDOW].unshift(stats());
+    }, bucketInterval);
     if (typeof interval.unref === 'function') interval.unref();
   }
-}
 
-function reset (status) {
+  get window () {
+    return this[STATS_WINDOW].slice();
+  }
+
   /**
-   * The number of times the breaker's action has failed
+   * The number of times the action for this breaker executed successfully
+   * during the current statistical window.
    */
-  status.failures = 0;
+  get successes () {
+    return this[STATS_WINDOW][0].successes;
+  }
+
   /**
-   * The number of times a fallback function has been executed
+   * The number of times the breaker's action has failed
+   * during the current statistical window.
    */
-  status.fallbacks = 0;
+  get failures () {
+    return this[STATS_WINDOW][0].failures;
+  }
+
   /**
-   * The number of times the action for this breaker executed successfully
+   * The number of times a fallback function has been executed
+   * during the current statistical window.
    */
-  status.successes = 0;
+  get fallbacks () {
+    return this[STATS_WINDOW][0].fallbacks;
+  }
+
   /**
-   * The number of times this breaker been rejected because it was fired, but in the open state.
+   * The number of times during the current statistical window that
+   * this breaker been rejected because it was in the open state.
    */
-  status.rejects = 0;
+  get rejects () {
+    return this[STATS_WINDOW][0].rejects;
+  }
+
   /**
    * The number of times this circuit breaker has been fired
+   * during the current statistical window.
    */
-  status.fires = 0;
+  get fires () {
+    return this[STATS_WINDOW][0].fires;
+  }
+
   /**
    * The number of times this circuit breaker has timed out
+   * during the current statistical window.
    */
-  status.timeouts = 0;
+  get timeouts () {
+    return this[STATS_WINDOW][0].timeouts;
+  }
 }
 
+const stats = () => ({
```

Comment:
I will merge from master before pushing my changes. There will be conflicts in this file, but I will deal with them.